subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java

```diff
@@ -67,13 +67,13 @@ public T maybeCreate(String name) {
 
     @Override
     public T create(String name, Closure configureClosure) {
-        assertMutable("create(String, Closure)");
+        assertCanMutate("create(String, Closure)");
         return create(name, ConfigureUtil.configureUsing(configureClosure));
     }
 
     @Override
     public T create(String name, Action<? super T> configureAction) throws InvalidUserDataException {
-        assertMutable("create(String, Action)");
+        assertCanMutate("create(String, Action)");
```

Comment:
That would mean assertCanAdd would need to accept both the method name and the name of thing we are trying to add.

Really I think assertCanAdd is a bad name. I would expect that if I call `assertCanAdd` that means I can add something. To satisfy that assertion, as you mention, that should verify 
1. An object with that name is not present
2. The container is mutable

I think I would prefer to keep these operations separate, particularly since there are instances where `assertCanAdd` are currently called without an `assertCanMutate` (presumably since we've already checked that we can mutate elsewhere)

I would consider renaming this to something more accurate, like `assumeElementNotPresent` or similar

Reply:
Makes sense.